apps/frontend/src/store/data_filters.ts

```diff
@@ -353,18 +374,126 @@ export const FilteredDataModule = getModule(FilteredData);
     omit_overlayed_controls: f.omit_overlayed_controls || false,
     ...f
   };
   return JSON.stringify(newFilter);
 }
+
+export function filterByStatus(
+  filter: Filter,
+  controls: readonly context.ContextualizedControl[]
+): readonly context.ContextualizedControl[] {
+  if (filter.status && filter.status?.length !== 0) {
+    return controls.filter((control) =>
+      filter.status?.some((term) => {
+        if (term.toLowerCase() === 'waived') {
+          return control.hdf.waived;
+        } else {
+          return control.root.hdf.status
+            .toLowerCase()
+            .includes(term.toLowerCase());
+        }
+      })
+    );
+  } else {
+    return controls;
+  }
+}
+
+export function filterBySeverity(
+  filter: Filter,
+  controls: readonly context.ContextualizedControl[]
+): readonly context.ContextualizedControl[] {
+  if (filter.severity && filter.severity?.length !== 0) {
+    return controls.filter((control) =>
+      filter.severity?.some((term) =>
+        control.root.hdf.severity.toLowerCase().includes(term.toLowerCase())
+      )
+    );
+  } else {
+    return controls;
+  }
+}
+
+export function filterByControlID(
+  filter: Filter,
+  controls: readonly context.ContextualizedControl[]
+): readonly context.ContextualizedControl[] {
+  if (filter.ids && filter.ids?.length !== 0) {
+    return controls.filter((control) =>
+      filter.ids?.some((term) =>
+        control.hdf.wraps.id.toLowerCase().includes(term.toLowerCase())
+      )
+    );
+  } else {
+    return controls;
+  }
+}
+
+export function filterByTitle(
+  filter: Filter,
+  controls: readonly context.ContextualizedControl[]
+): readonly context.ContextualizedControl[] {
+  if (filter.titleSearchTerms && filter.titleSearchTerms?.length !== 0) {
+    return controls.filter((control) =>
+      filter.titleSearchTerms?.some((term) =>
+        control.hdf.wraps.title?.toLowerCase().includes(term.toLowerCase())
+      )
+    );
+  } else {
+    return controls;
+  }
+}
+
+export function filterByDescription(
+  filter: Filter,
+  controls: readonly context.ContextualizedControl[]
+): readonly context.ContextualizedControl[] {
+  if (
+    filter.descriptionSearchTerms &&
+    filter.descriptionSearchTerms?.length !== 0
+  ) {
+    return controls.filter((control) =>
+      filter.descriptionSearchTerms?.some((term) =>
+        control.hdf.wraps.desc?.toLowerCase().includes(term.toLowerCase())
+      )
+    );
+  } else {
+    return controls;
+  }
+}
+
+export function filterByNIST(
+  filter: Filter,
+  controls: readonly context.ContextualizedControl[]
+): readonly context.ContextualizedControl[] {
+  if (filter.cciIdFilter && filter.cciIdFilter?.length !== 0) {
+    return controls.filter((control) => {
+      return control.hdf.raw_nist_tags?.some((tag) =>
+        filter.cciIdFilter?.some((term) =>
+          tag.toLowerCase().includes(term.toLowerCase())
+        )
+      );
+    });
+  } else {
+    return controls;
+  }
+}
```

Comment:
All of these functions are basically exactly the same, can you create a helper function that takes 3 parameters, the item to filter and the array result to call it on, and just call it for all of these?

Basically the body of this could become

```javascript
return filter_controls_by('full_code', controls, filter.codeSearchTerms)
```